### PR TITLE
[DebugInfo] Handle dbg.assigns in FastISel

### PR DESCRIPTION
**Pull request by jmorse**
There are some rare circumstances where dbg.assign intrinsics can reach FastISel. They are a more specialised kind of dbg.value intrinsic with more information about the originating alloca. They only occur during optimisation, but might reach FastISel through always_inlining an optimised function into an optnone function.

This is a slight problem as it's not safe (for debug-info accuracy) to ignore any intrinsics, and for RemoveDIs (the intrinsic-replacement project) it causes a crash through an unhandled switch case. To get around this, we can just treat the dbg.assign as a dbg.value (it's an actual subclass) and use the variable location information from the dbg.value fields. This loses a small amount of debug-info about stack locations, but is more accurate than just ignoring the intrinsic.

(This has popped up deep in an LTO build of a large codebase while testing RemoveDIs, I figured it'd be good to fix it for the intrinsic-form at the same time, just to demonstrate the correct behaviour).

### CODE REVIEW

File: llvm/test/DebugInfo/X86/dbg-assign-fastisel.ll [hunk 1]
```diff
@@ -0,0 +1,44 @@
+; RUN: llc %s -fast-isel=true -start-after=codegenprepare -stop-before=finalize-isel -o - | FileCheck %s
+; RUN: llc %s -fast-isel=true -start-after=codegenprepare -stop-before=finalize-isel -o - --try-experimental-debuginfo-iterators | FileCheck %s
```
Comment (arsenm) [inline llvm/test/DebugInfo/X86/dbg-assign-fastisel.ll:2]: don't really need the =true?

**Comment by jmorse** [comment]
For SelectionDAG it's already had dbg.assign/"assignment-tracking" plumbed into it. However it looks like GlobalISel could do with this too -- I'll add that to this review.

**Review by SLTozer — APPROVED** [review]
Unless other reviewers have any more comments to add, this change LGTM.

**Comment by jmorse** [comment]
In a slight rush as we can't keep-up with the main branch forever while waiting to turn RemoveDIs on; happy to revise and extend this patch in the future if it's needed.

**Pull request merged**